crates/api_crud/src/user/create.rs

```diff
@@ -238,3 +227,319 @@ pub async fn register(
+
+    // prevent registration if registration is closed
+    if local_site.registration_mode == RegistrationMode::Closed {
+      Err(LemmyErrorType::RegistrationClosed)?
+    }
+
+    // prevent registration if registration is closed for OAUTH providers
+    if !local_site.oauth_registration {
+      return Err(LemmyErrorType::OauthRegistrationClosed)?;
+    }
+
+    // Extract the OAUTH email claim from the returned user_info
+    let email: String;
+    if let Some(email_value) = user_info.get("email") {
+      email = serde_json::from_value::<String>(email_value.clone())
+        .ok()
+        .ok_or(LemmyErrorType::OauthLoginFailed)?;
+    } else {
+      return Err(LemmyErrorType::OauthLoginFailed)?;
+    }
+
+    // Lookup user by OAUTH email and link accounts
+    local_user_view = LocalUserView::find_by_email(&mut context.pool(), &email).await?;
+
+    let person;
+    if let Some(user_view) = local_user_view {
+      // user found by email => link and login if linking is allowed
+
+      if oauth_provider.account_linking_enabled {
+        // Link with OAUTH => Login user
+        let oauth_account_form =
+          OAuthAccountInsertForm::new(user_view.local_user.id, oauth_provider.id, oauth_user_id);
+
+        OAuthAccount::create(&mut context.pool(), &oauth_account_form)
+          .await
+          .ok()
+          .ok_or(LemmyErrorType::OauthLoginFailed)?;
+
+        local_user = user_view.local_user.clone();
+
+        check_user_valid(&user_view.person)?;
+        check_registration_application(&user_view, &site_view.local_site, &mut context.pool())
+          .await?;
+        check_email_verified(&user_view, &site_view)?;
+      } else {
+        // email already registered
+        return Err(LemmyErrorType::EmailAlreadyExists)?;
+      }
+    } else {
+      // No user was found by email => Register as new user
+
+      // Extract the OAUTH name claim from the returned user_info
+      let user_name: String;
+      if let Some(user_name_value) = user_info.get(oauth_provider.name_claim) {
+        user_name = serde_json::from_value::<String>(user_name_value.clone())
+          .ok()
+          .ok_or(LemmyErrorType::OauthLoginFailed)?;
+      } else {
+        return Err(LemmyErrorType::OauthLoginFailed)?;
+      }
+
+      // remove spaces from username
+      let username = str::replace(&user_name, " ", "_");
+      let username = username + &rand::thread_rng().gen_range(0..999).to_string();
+
+      // check for slurs
+      let slur_regex = local_site_to_slur_regex(&local_site);
+      check_slurs(&username, &slur_regex)
+        .ok()
+        .ok_or(LemmyErrorType::OauthLoginFailed)?;
+
+      let require_registration_application =
+        local_site.registration_mode == RegistrationMode::RequireApplication;
+
+      if local_site.site_setup && require_registration_application && data.answer.is_none() {
+        Err(LemmyErrorType::RegistrationApplicationAnswerRequired)?
+      }
```

Comment:
This is copy-pasted from `register()`, move it into a helper function to avoid duplicate code. Together with some of the other checks like registration code.